backend/cmm_helpers.ml

```diff
@@ -3313,6 +3321,71 @@ let prefetch_offset ~is_write locality (arg1, arg2) dbg =
 
+let native_pointer_cas size (arg1, arg2, arg3) dbg =
+  let op = Catomic { op = Compare_and_swap; size } in
+  if_operation_supported op ~f:(fun () ->
+      bind "set_to" arg3 (fun set_to ->
+          bind "compare_with" arg2 (fun compare_with ->
+              bind "dst" arg1 (fun dst ->
+                  tag_int (Cop (op, [compare_with; set_to; dst], dbg)) dbg))))
+
+let ext_pointer_cas size (arg1, arg2, arg3) dbg =
+  let op = Catomic { op = Compare_and_swap; size } in
+  if_operation_supported op ~f:(fun () ->
+      bind "set_to" arg3 (fun set_to ->
+          bind "compare_with" arg2 (fun compare_with ->
+              bind "dst" (int_as_pointer arg1 dbg) (fun dst ->
+                  tag_int (Cop (op, [compare_with; set_to; dst], dbg)) dbg))))
+
+let bigstring_cas size (arg1, arg2, arg3, arg4) dbg =
+  let op = Catomic { op = Compare_and_swap; size } in
+  if_operation_supported op ~f:(fun () ->
+      bind "set_to" arg4 (fun set_to ->
+          bind "compare_with" arg3 (fun compare_with ->
+              bind "idx" arg2 (fun idx ->
+                  bind "bs" arg1 (fun bs ->
+                      bind "bs_data"
+                        (Cop
+                           ( Cload (Word_int, Mutable),
+                             [field_address bs 1 dbg],
+                             dbg ))
+                        (fun bs_data ->
+                          bind "dst" (add_int bs_data idx dbg) (fun dst ->
+                              tag_int
+                                (Cop (op, [compare_with; set_to; dst], dbg))
+                                dbg)))))))
+
+let native_pointer_atomic add_or_sub size (arg1, arg2) dbg =
+  let op = Catomic { op = Fetch_and_add; size } in
+  if_operation_supported op ~f:(fun () ->
+      bind "src"
+        (match add_or_sub with `Add -> arg2 | `Sub -> neg_int arg2 dbg)
+        (fun src -> bind "dst" arg1 (fun dst -> Cop (op, [src; dst], dbg))))
+
+let ext_pointer_atomic add_or_sub size (arg1, arg2) dbg =
+  let op = Catomic { op = Fetch_and_add; size } in
+  if_operation_supported op ~f:(fun () ->
+      bind "src"
+        (match add_or_sub with `Add -> arg2 | `Sub -> neg_int arg2 dbg)
+        (fun src ->
+          bind "dst" (int_as_pointer arg1 dbg) (fun dst ->
+              Cop (op, [src; dst], dbg))))
+
+let bigstring_atomic add_or_sub size (arg1, arg2, arg3) dbg =
+  let op = Catomic { op = Fetch_and_add; size } in
+  if_operation_supported op ~f:(fun () ->
+      bind "src"
+        (match add_or_sub with `Add -> arg3 | `Sub -> neg_int arg3 dbg)
+        (fun src ->
+          bind "idx" arg2 (fun idx ->
+              bind "bs" arg1 (fun bs ->
+                  bind "bs_data"
+                    (Cop
+                       (Cload (Word_int, Mutable), [field_address bs 1 dbg], dbg))
+                    (fun bs_data ->
+                      bind "dst" (add_int bs_data idx dbg) (fun dst ->
+                          Cop (op, [src; dst], dbg)))))))
+
```

Comment:
Can you try to factor out some of the duplication into helper functions, hopefully voiding  polymorphic variants?

Reply:
Done